### PR TITLE
[Tests] Added tests for all the optimizers

### PR DESCRIPTION
**Pull request by Lucaspbordignon**
Just added tests for all the implemented optimizers.

### CODE REVIEW

File: test/test_optimizers.h [hunk 1]
```diff
@@ -4,6 +4,106 @@
...
+    vec_t weights = {.2, .4, .006, -.77, -.010};
+    vec_t gradients = {1.0, -3.24, -.6, 2.79, 1.82};
+
+    // Defining the expected updates
+    vec_t first_update = {0.1900, 0.4100, 0.0160, -0.7800, -0.0200};
+    vec_t second_update = {0.1829, 0.4170, 0.0230, -0.7870, -0.0270};
+
+    // Testing
+    optimizer.update(gradients, weights, false);
+    for (auto i = 0u; i < weights.size(); i++)
```
Comment (karandesai-96) [inline test/test_optimizers.h]: Just for the sake of consistency with other parts of code, could you change this to `size_t i = 0` ? Same for all for loops.

**Comment by karandesai-96** [comment]
Oops, my laptop's touchpad has got greasy, this got approved before time by mistake. @Lucaspbordignon 
 please go ahead and make a modification.

**Comment by Lucaspbordignon** [comment]
@karandesai-96 Modifications made!

**Comment by karandesai-96** [comment]
Thanks @Lucaspbordignon :smile:

File: test/test_optimizers.h [hunk 2]
```diff
@@ -4,6 +4,106 @@
 
 namespace tiny_dnn {
 
+TEST(optimizers, adagrad_update) {
+    adagrad optimizer;
+
+    vec_t weights = {.2, .4, .006, -.77, -.010};
```
Comment (edgarriba) [inline test/test_optimizers.h]: for consistency: `.2, .4, .006, -.77, -.010 -> 0.2, 0.4, 0.006, -0.77, -0.010`

Comment (edgarriba) [inline test/test_optimizers.h]: for consistency: .2, .4, .006, -.77, -.010 -> 0.2, 0.4, 0.006, -0.77, -0.010

File: test/test_optimizers.h [hunk 3]
```diff
@@ -4,6 +4,106 @@
...
+    vec_t gradients = {1.0, -3.24, -.6, 2.79, 1.82};
+
+    // Defining the expected updates
+    vec_t first_update = {0.1900, 0.4100, 0.0160, -0.7800, -0.0200};
+    vec_t second_update = {0.1829, 0.4170, 0.0230, -0.7870, -0.0270};
+
+    // Testing
+    optimizer.update(gradients, weights, false);
+    for (size_t i = 0; i < weights.size(); i++)
+        EXPECT_NEAR(first_update[i], weights[i], 1e-3);
```
Comment (edgarriba) [inline test/test_optimizers.h]: use brackets in if/else statements

Comment (edgarriba) [inline test/test_optimizers.h]: precision should be higher, did you try `1e-6` ?

**Comment by edgarriba** [comment]
@Lucaspbordignon could you also add smoke tests for constructors and member variables ?

**Comment by edgarriba** [comment]
@Lucaspbordignon thanks for the contribution !

**Comment by edgarriba** [comment]
in case you want to follow contributing, visit us in gitter developers. Discussions happen there and probably there more stuff to do related to optimization. Thanks again !

**Pull request merged**